src/compiler/checker.ts

```diff
@@ -6154,15 +6154,21 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
                 const type = instantiateType(getWidenedType(getRegularTypeOfExpression(expr)), context.mapper);
                 return typeToTypeNodeHelper(type, context);
             },
-            serializeTypeOfDeclaration(syntacticContext, declaration, symbol) {
+            serializeTypeOfDeclaration(syntacticContext, declaration, symbol, widen) {
                 // Get type of the symbol if this is the valid symbol otherwise get type at location
                 const context = syntacticContext as NodeBuilderContext;
                 symbol ??= getSymbolOfDeclaration(declaration);
                 let type = context.enclosingSymbolTypes?.get(getSymbolId(symbol));
                 if (type === undefined) {
-                    type = symbol && !(symbol.flags & (SymbolFlags.TypeLiteral | SymbolFlags.Signature))
-                        ? instantiateType(getWidenedLiteralType(getTypeOfSymbol(symbol)), context.mapper)
-                        : errorType;
+                    const symbolType = getTypeOfSymbol(symbol);
+                    if (widen) {
+                        type = instantiateType(getWidenedType(symbolType), context.mapper);
+                    }
+                    else {
+                        type = symbol && !(symbol.flags & (SymbolFlags.TypeLiteral | SymbolFlags.Signature))
+                            ? instantiateType(getWidenedLiteralType(symbolType), context.mapper)
+                            : errorType;
+                    }

```

Comment:
Seems like this could be just:

```suggestion
                    type = widen || (symbol && !(symbol.flags & (SymbolFlags.TypeLiteral | SymbolFlags.Signature)))
                        ? instantiateType(getWidenedLiteralType(getTypeOfSymbol(symbol)), context.mapper)
                        : errorType;
```

Reply:
(the original code with one more expression)

Reply:
The problem is that one does just literal widening and the new branch widens any type.

Reply:
Oh, did not notice that, oops. I am surprised that `getWidenedType` doesn't just do that.